File: card-ui/src/main/java/com/adyen/checkout/card/ui/CardNumberInput.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by arman on 19/7/2019.
 */

package com.adyen.checkout.card.ui;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.util.AttributeSet;

public class CardNumberInput extends NumberInputEditText {

    private static final int MAX_DIGIT_LENGTH = 19;
    private static final int MAX_DIGIT_SEPARATOR_COUNT = 4;
    private static final char DIGIT_SEPARATOR = ' ';

    public CardNumberInput(@NonNull Context context) {
        this(context, null);
    }

    public CardNumberInput(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CardNumberInput(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public int maxLength() {
        return MAX_DIGIT_LENGTH + MAX_DIGIT_SEPARATOR_COUNT;
    }

    @NonNull
    @Override
    public String getRawValue() {
        return getText().toString().replace(DIGIT_SEPARATOR + "", "");
    }

    @Override
    protected void afterTextChanged(@NonNull Editable editable) {
        final String initial = editable.toString();
        String processed = initial.trim();
        processed = processed.replaceAll("(\\d{4})(?=\\d)", "$1" + DIGIT_SEPARATOR);
        if (!initial.equals(processed)) {
            editable.replace(0, initial.length(), processed);
        }

        super.afterTextChanged(editable);
    }
}

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberInput.kt
/*
 * Copyright (c) 2022 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 25/3/2022.
 */

package com.adyen.checkout.card.internal.ui.view

import android.content.Context
import android.os.Build
import android.text.Editable
import android.text.InputType
import android.text.method.DigitsKeyListener
import android.util.AttributeSet
import androidx.annotation.RestrictTo
import com.adyen.checkout.card.internal.util.CardNumberUtils
import com.adyen.checkout.core.old.ui.validation.CardNumberValidator
import com.adyen.checkout.ui.core.old.internal.ui.view.AdyenTextInputEditText

/**
 * Input that support formatting for card number.
 */
class CardNumberInput
@JvmOverloads
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AdyenTextInputEditText(context, attrs, defStyleAttr) {

    private var isAmexCard = false
    override val rawValue: String
        get() = text.toString().replace(DIGIT_SEPARATOR, "")

    init {
        enforceMaxInputLength(CardNumberValidator.MAXIMUM_CARD_NUMBER_LENGTH + MAX_DIGIT_SEPARATOR_COUNT)
        inputType = InputType.TYPE_CLASS_NUMBER
        keyListener = DigitsKeyListener.getInstance(SUPPORTED_DIGITS + DIGIT_SEPARATOR)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            setAutofillHints(AUTOFILL_HINT_CREDIT_CARD_NUMBER)
        }
    }

    /**
     * Enable Amex formatting.
     */
    fun setAmexCardFormat(value: Boolean) {
        // first time detecting is Amex card
        if (!isAmexCard && value) {
            isAmexCard = true
            afterTextChanged(editableText)
            return
        }
        isAmexCard = value
    }

    override fun afterTextChanged(editable: Editable) {
        val initial = editable.toString()
        var processed = initial.replace(DIGIT_SEPARATOR, "")
        processed = formatProcessedString(processed)
        if (initial != processed) {
            editable.replace(0, initial.length, processed)
        }
        super.afterTextChanged(editable)
    }

    private fun formatProcessedString(unformattedString: String): String {
        return CardNumberUtils.formatCardNumber(
            unformattedString = unformattedString,
            maskPartsLengths = if (isAmexCard) AMEX_CARD_NUMBER_MASK else DEFAULT_CARD_NUMBER_MASK,
            separator = DIGIT_SEPARATOR,
        )
    }

    companion object {
        private const val MAX_DIGIT_SEPARATOR_COUNT = 4
        private const val DIGIT_SEPARATOR = " "
        private const val SUPPORTED_DIGITS = "0123456789"

        // 4 characters then 6 then 5 then 4. Example: 1234 123456 12345 1234
        private val AMEX_CARD_NUMBER_MASK = listOf(4, 6, 5, 4)

        // 4 characters then 4 then 4 then 4 then 3. Example: 1234 1234 1234 1234 123
        private val DEFAULT_CARD_NUMBER_MASK = listOf(4, 4, 4, 4, 3)
    }
}
